src/main/java/org/magicdgs/readtools/utils/distmap/DistmapGATKWriter.java

```diff
@@ -0,0 +1,146 @@
+     * @param sourceName the name for the source where we are writing to.
+     */
+    public DistmapGATKWriter(final PrintStream writer, final String sourceName,
+            final boolean paired) {
+        this.writer = writer;
+        this.sourceName = sourceName;
+        this.singleEndHandler = (paired)
+                ? new DistmapPairedConsumer(this::addPair)
+                : read -> printAndCheckError(DistmapEncoder.encode(read));
+    }
+
+    /**
+     * Adds a read to the writer. If added in pair-end mode, the reads in the pair are expected to
+     * be passed one after the other.
+     */
+    @Override
+    public void addRead(final GATKRead read) {
+        singleEndHandler.accept(read);
+    }
+
+    /**
+     * Writes a pair of reads in the Distmap format.
+     *
+     * @param pair the pair of reads.
+     */
+    public void addPair(final Tuple2<GATKRead, GATKRead> pair) {
+        printAndCheckError(DistmapEncoder.encode(pair));
+    }
+
+    /**
+     * Close the underlying writer.
+     *
+     * @throws IllegalStateException if the writer is in paired and not second read is added.
+     */
+    @Override
+    public void close() throws IOException {
+        // close the handler if it is a DistmapPairedConsumer
+        if (singleEndHandler instanceof Closeable) {
+            ((Closeable) singleEndHandler).close();
+        }
+        writer.close();
+    }
+
+    // helper method to print and with the writer and check if an error occurs to throw an exception
+    private void printAndCheckError(final String toPrint) {
+        writer.println(toPrint);
+        if (writer.checkError()) {
+            throw new UserException.CouldNotCreateOutputFile(sourceName,
+                    "writing error",
+                    new IOException());
+        }
+    }
+
+
+    // private class to handle pair-end reads
+    // it keeps the first and second in pair
+    // it is closeable because it should throw if the consumer have an unexpected ending
+    private final static class DistmapPairedConsumer implements Consumer<GATKRead>, Closeable {
+
+        // first and second read passed to the handler, afterwards they are cleaned
+        private GATKRead first = null;
+
+        private final Consumer<Tuple2<GATKRead, GATKRead>> consumer;
+
+        private DistmapPairedConsumer(final Consumer<Tuple2<GATKRead, GATKRead>> consumer) {
+            this.consumer = consumer;
+        }
+
+        @Override
+        public void accept(final GATKRead read) {
+            if (first == null) {
+                first = read;
+            } else {
+                // assume that it is the second
+                consumer.accept(new Tuple2<>(first, read));
+                first = null;
+            }
+        }
+
+        // throws if the first is not null
+        @Override
+        public void close() throws IOException {
+            if (first != null) {
+                throw new IOException(
```

Comment:
I'm still not sure about this exception. It is an exception for output, but due to an illegal state...so it could come from a bad implementation or an user error (broken input). Anyway, this will blow at some point before, but maybe will be better to make it catchable in a different way...